File: lib/features/notes/domain/repositories/notes_repository.dart
import 'package:dartz/dartz.dart';
import '../../../../core/error/failures.dart';
import '../entities/note.dart';

abstract class NotesRepository {
  Future<Either<Failure, List<Note>>> getAllNotes();
  Future<Either<Failure, void>> addNote(Note note);
  Future<Either<Failure, void>> deleteNote(int id);
}

File: android/app/src/main/kotlin/com/example/flutter_clean_arch_notes/feature_note/domain/repository/NotesRepository.kt
package com.example.flutter_clean_arch_notes.feature_note.domain.repository

import kotlinx.coroutines.flow.Flow
import com.example.flutter_clean_arch_notes.feature_note.domain.model.Note

/**
 * NotesRepository defines the abstract contract for
 * all data operations related to notes.
 *
 * The Domain layer depends on this interface, not on
 * any concrete data implementation (like Room or Retrofit).
 */
interface NotesRepository {

    /**
     * Fetch all notes from the data source.
     * Returns a Flow that emits updates whenever data changes.
     */
    fun getAllNotes(): Flow<List<Note>>

    /**
     * Add or update a note in the data source.
     */
    suspend fun addNote(note: Note)

    /**
     * Delete a note by its ID.
     */
    suspend fun deleteNote(noteId: Int)
}
